modules/lang-painless/src/main/java/org/elasticsearch/painless/node/EChain.java

```diff
@@ -233,6 +248,11 @@ private void analyzeWrite(final CompilerSettings settings, final Definition defi
     private void analyzeRead() {
         final ALink last = links.get(links.size() - 1);
 
+        // If the load node is a DEF node, we adapt its after type to use _this_ expected output type:
+        if (last instanceof ADefLink && this.expected != null) {
+            last.after = this.expected;
+        }
+
```

Comment:
This is close, but I think this needs to, unfortunately, be placed in an overridden cast method.  this.expected isn't guaranteed to be set until after the analyze phase because of possible promotion.

Something like the following:
```
    @Override
    AExpression cast(final CompilerSettings settings, final Definition definition, final Variables variables) {
        final ALink last = links.get(links.size() - 1);

        // If the load node is a DEF node, we adapt its after type to use _this_ expected output type.  
        // Note we avoid doing this if the cast is explicit because we want to ensure that the cast
        // will do primitive narrowing operations and upcasts at runtime without throwing an 
        // exception.  We also avoid doing this if a store is being done since the value from
        // the store is what will be on the stack.
        if (!explicit && expression == null && last instanceof ADefLink) {
            last.after = this.expected;
            this.actual = this.expected;

            return this;
        } else {
            return super.cast(settings, definition, variables);
        }
    }
```